pygbm/splitting.py

```diff
@@ -60,78 +61,163 @@ def __init__(self, n_features, binned_features, n_bins,
+
+    left_cnt = np.empty(n_threads, dtype=np.int32)
+    right_cnt = np.empty(n_threads, dtype=np.int32)
+
+    # map indices to tmp_left and tmp_right
+    for thread_i in prange(n_threads):
+        left_count = 0
+        right_count = 0
+
+        start = starts[thread_i]
+        stop = start + sizes[thread_i]
+        for i in range(start, stop):
+            sample_idx = sample_indices[i]
             if binned_feature[sample_idx] <= split_info.bin_idx:
-                sample_indices_left[left_idx] = sample_idx
-                left_idx += 1
+                tmp_left[start + left_count] = sample_idx
+                left_count += 1
             else:
-                sample_indices_right[right_idx] = sample_idx
-                right_idx += 1
+                tmp_right[start + right_count] = sample_idx
+                right_count += 1
 
-        return sample_indices_left[:left_idx], sample_indices_right[:right_idx]
+        left_cnt[thread_i] = left_count
+        right_cnt[thread_i] = right_count
 
-    def find_node_split(self, sample_indices):
-        loss_dtype = self.all_gradients.dtype
+    left_write_offset = np.zeros(n_threads, dtype=np.int32)
+    right_write_offset = np.zeros(n_threads, dtype=np.int32)
+    left_write_offset[1:] = np.cumsum(left_cnt[:-1])
+    right_write_offset[1:] = np.cumsum(right_cnt[:-1])
 
-        if sample_indices.shape[0] == self.all_gradients.shape[0]:
-            # Root node: the ordering of sample_indices and all_gradients
-            # are expected to be consistent in this case.
-            ordered_gradients = self.all_gradients
-            ordered_hessians = self.all_hessians
-        else:
-            ordered_gradients = np.empty_like(sample_indices, dtype=loss_dtype)
-            if self.constant_hessian:
-                ordered_hessians = self.all_hessians
-                for i, sample_idx in enumerate(sample_indices):
-                    ordered_gradients[i] = self.all_gradients[sample_idx]
-            else:
-                ordered_hessians = np.empty_like(sample_indices,
-                                                 dtype=loss_dtype)
-                for i, sample_idx in enumerate(sample_indices):
-                    ordered_gradients[i] = self.all_gradients[sample_idx]
-                    ordered_hessians[i] = self.all_hessians[sample_idx]
+    # put tmp_left and tmp_right into samples_indices_left/right
+    for thread_i in prange(n_threads):
 
-        sum_gradients = ordered_gradients.sum()
-        if self.constant_hessian:
-            n_samples = sample_indices.shape[0]
-            sum_hessians = self.constant_hessian_value * float32(n_samples)
-        else:
-            sum_hessians = ordered_hessians.sum()
+        start, stop = starts[thread_i], starts[thread_i] + left_cnt[thread_i]
+        for i, sample_idx in enumerate(range(start, stop)):
+            sample_indices_left[left_write_offset[thread_i] + i] = \
+                tmp_left[sample_idx]
```

Comment:
The `sample_idx` variable name here is confusing. This integer is not an index of a sample in the original training set feature array. It could be renamed to something more neutral such as j.

Alternatively, I wonder if numba accepts something like:

```python
sample_indices_left[left_write_offset[thread_i]:left_write_offset[thread_i]+left_cnt[thread_i]] = tmp_left[start:stop]
```

Off-course the above is ugly and should be rewritten with additional temporary index variables to make it more readable but you get the idea.

Reply:
In the future, if we want to reuse the `sample_indices_left`/`sample_indices_right` buffers for consecutive splits in the same tree we could always change the returned values to return copies to be stored on the nodes that result from the split:

```python
    return (context.sample_indices_left[:total_left_count].copy(),
               context.sample_indices_right[:total_right_count].copy())
```

Reply:
> we could always change the returned values to return copies

Yes, but that wouldn't save any time / memory right? The copied array would still need to be allocated.

Let me try to explain what LightGBM is doing to avoid those allocations at every node:

To keep track of how the samples are dispatched over the leaves of the tree, Lightgbm uses a `DataPartition` object. There's a one-to-one relationship between a grower and a data_partition. The main structure of this `DataPartition` is a `indices_` array which looks like this:

`[qwert|yuio|pa|sdfghjk|lzxc]`

which basically indicates that the samples `qwert` are in the first leaf, `yuio` in the second, etc.

If the first leaf is split, the array would become, e.g.

`[qt|erw|yuio|pa|sdfghjk|lzxc]` (the order of the samples inside a leaf is irrelevant)
This is why they need the `tmp...` array in the `split_indices` method BTW, they can't work directly on `indices_` because they would overwrite some parts of it.

Reply:
For slices:

Tests pass with `parallel=False` but not with `parallel=True`. Not sure if this is a numba bug because I could not reproduce it. I'll keep the current version for now.

```python
        start_tmp = starts[thread_idx]
        stop_tmp = starts[thread_idx] + left_counts[thread_idx]
        start = left_write_offset[thread_idx]
        stop = left_write_offset[thread_idx] + left_counts[thread_idx]

        sample_indices_left[start:stop] = tmp_left[start_tmp:stop_tmp]
```

Reply:
Alright. Thanks for the explanation.